### PR TITLE
Handle completions in interface / type literal similar to class

### LINKED ISSUE #22672: Auto completion for interface method should not automatically add an import
**TypeScript Version:**  2.7.2, 2.8.0-dev.20180317

**Search Terms:** automatic import interface method

When writing an interface with overloaded methods, using names that exist in some NodeJS core module causes the auto completion to contain entries from that module.

**Code**

```ts
export interface Foo {
	removeListener(event: "foo", callback: () => void): this;
	removeLi // BUG HERE!
}
```

**Expected behavior:**
Either no auto completion at all or auto completion for the overloaded method name `removeListener` from the interface `Foo`

**Actual behavior:**
The auto completion contains 3 methods from the `cluster` module with an auto import quick fix:
import
Pressing TAB causes my code to end up with an unnecessary import.

### PR DESCRIPTION
**Pull request by andy-ms**
Fixes #22672

Previously we provided global completions inside of interfaces / type literals. Now we will only provide the `readonly` keyword.

### CODE REVIEW

File: src/services/completions.ts [hunk 1]
```diff
@@ -2282,4 +2222,33 @@ namespace ts.Completions {
...
+    function tryGetObjectTypeDeclarationCompletionContainer(contextToken: Node, location: Node): ObjectTypeDeclaration | undefined {
+        // class c { method() { } | method2() { } }
+        if (location.kind === SyntaxKind.SyntaxList && isObjectTypeDeclaration(location.parent)) return location.parent;
+        switch (contextToken.kind) {
+            case SyntaxKind.SemicolonToken: // class c {getValue(): number; | }
+            case SyntaxKind.CloseBraceToken: // class c { method() { } | }
+                // class c { method() { } b| }
+                return isFromObjectTypeDeclaration(location) && (location.parent as ClassElement | TypeElement).name === location
+                    ? location.parent.parent as ObjectTypeDeclaration
+                    : tryCast(location, isClassLike);
```
Comment (sheetalkamat) [inline src/services/completions.ts]: why is this not ```isObjectTypeDeclaration``` ?

File: src/services/completions.ts [hunk 2]
```diff
@@ -2282,4 +2222,33 @@ namespace ts.Completions {
...
+            case SyntaxKind.OpenBraceToken: // class c { |
+            case SyntaxKind.CommaToken: // class c {getValue(): number, | }
+                return tryCast(contextToken.parent, isObjectTypeDeclaration);
+            default:
+                return isFromObjectTypeDeclaration(contextToken) &&
+                    (isClassMemberCompletionKeyword(contextToken.kind) || isIdentifier(contextToken) && isClassMemberCompletionKeywordText(contextToken.text))
```
Comment (sheetalkamat) [inline src/services/completions.ts]: hmm dont you need it to be classMemberCompletion if its call location and "readonly" in type literals and interfaces

File: tests/cases/fourslash/completionEntryForClassMembers.ts [hunk 3]
```diff
@@ -114,7 +114,7 @@
 ////}
 ////class O extends B {
 ////    constructor(public a) {
-////    },
```
Comment (sheetalkamat) [inline tests/cases/fourslash/completionEntryForClassMembers.ts]: why is this change?

File: tests/cases/fourslash/completionListAtIdentifierDefinitionLocations_interfaceMembers.ts [hunk 4]
```diff
@@ -1,7 +0,0 @@
-/// <reference path='fourslash.ts' />
-
-////var aa = 1;
-
-////interface a { /*interfaceValue1*/
-
-goTo.eachMarker(() => verify.completionListIsEmpty());
```
Comment (sheetalkamat) [inline tests/cases/fourslash/completionListAtIdentifierDefinitionLocations_interfaceMembers.ts]: can you instead of deleting this have test case that verifies "aa" is not present in the completion

File: tests/cases/fourslash/completionListAtIdentifierDefinitionLocations_interfaceMembers2.ts [hunk 5]
```diff
@@ -1,7 +0,0 @@
-/// <reference path='fourslash.ts' />
-
-////var aa = 1;
-
-////interface a { f/*interfaceValue2*/
-
-goTo.eachMarker(() => verify.completionListIsEmpty());
```
Comment (sheetalkamat) [inline tests/cases/fourslash/completionListAtIdentifierDefinitionLocations_interfaceMembers2.ts]: can you instead of deleting this have test case that verifies "aa" is not present in the completion

File: tests/cases/fourslash/completionsInterfaceElement.ts [hunk 6]
```diff
@@ -0,0 +1,11 @@
+/// <reference path="fourslash.ts" />
+
+////const foo = 0;
+////interface I {
+////    m(): void;
+////    fo/*i*/
+////}
+////type T = { fo/*t*/ };
```
Comment (sheetalkamat) [inline tests/cases/fourslash/completionsInterfaceElement.ts]: Also need test case when interface I { /*marker*/ } and similar type literal

File: tests/cases/fourslash/completionsInterfaceElement.ts [hunk 7]
```diff
@@ -0,0 +1,11 @@
...
+
+//verify.completionsAt("i", ["readonly"]);
```
Comment (sheetalkamat) [inline tests/cases/fourslash/completionsInterfaceElement.ts]: why is this commented out?

File: src/services/completions.ts [hunk 8]
```diff
@@ -1527,58 +1528,51 @@ namespace ts.Completions {
          * Relevant symbols are stored in the captured 'symbols' variable.
          */
         function tryGetClassLikeCompletionSymbols(): GlobalsSearch {
-            const classLikeDeclaration = tryGetClassLikeCompletionContainer(contextToken);
-            if (!classLikeDeclaration) return GlobalsSearch.Continue;
+            const decl = contextToken && tryGetObjectTypeDeclarationCompletionContainer(contextToken, location);
```
Comment (sheetalkamat) [inline src/services/completions.ts]: this seems like change from previous behaviour where event if contextToken was not present we would use location to decide on class location. Why is this changed ?

File: tests/cases/fourslash/completionListAtIdentifierDefinitionLocations_interfaceMembers.ts (hunk 4 above)
Comment (andy-ms) [reply tests/cases/fourslash/completionListAtIdentifierDefinitionLocations_interfaceMembers.ts]: `completionsInterfaceElement` does basically the same thing -- `verify.completionsAt` will fail if there were completions we didn't list.

File: tests/cases/fourslash/completionListAtIdentifierDefinitionLocations_interfaceMembers2.ts (hunk 5 above)
Comment (andy-ms) [reply tests/cases/fourslash/completionListAtIdentifierDefinitionLocations_interfaceMembers2.ts]: There is a test in `completionsInterfaceElement` just like this.
